### ISSUE TITLE
Update PageSpeed Insights to be active by default for new installs

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#4133]
## Feature Description

As per the [Unified Dashboard Design](https://docs.google.com/document/d/1StGx76XXisE9MfwM1916D9fn7-H9BRiRgizRejTsecI/edit?pli=1#heading=h.t2gbzjvuj834), the PageSpeed Insights should be an always-activated/auto-activated module that can't be disabled.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

**Despite this technically being part of the _Unified dashboard_ epic, this can be released decoupled, without the feature flag, i.e. as soon as it is ready.**

* The PSI module should be activated by default for new users (or users that reset and then re-setup the plugin). In other words, it should really just be activated per the default value of the option, which would mean existing sites that already have the option set won't be affected.
* The CTA versions of the PSI widget on the dashboard should be removed (since once users consciously deactivate it, we shouldn't prompt them to activate it again).

### Previous ACs (outdated)

* The PageSpeed Insights Module should be changed to be a "force-active" module, i.e. it should always be active regardless of the active modules option, and it should not allow to be disabled.
* The CTA to activate PageSpeed Insights should be removed, since that state will no longer be relevant. The PageSpeed widget can be modified to render nothing if PageSpeed Insights is inactive (which is not realistically possible, but should still be covered here, similar to how we do for Search Console widgets).
* The special conditions for displaying alternate content when having successfully activated `pagespeed-insights` in the success notification `DashboardSetupAlerts` should be removed, since it will no longer be possible to land there.
* A new tooltip tour `pagespeedInsightsForceActive` should be added, to inform users of sites that didn't have the module activated before about that it's now active (it should not be shown on sites where the module was already active).
    * It should show if on the Site Kit dashboard.
    * It should show if the initial Site Kit version is older than whatever release version this issue gets planned for.
    * It should show (using `checkRequirements`) if the site **does not** have `pagespeed-insights` in the list of active modules from the option. Note that this somehow needs to look at the option value explicitly, since when checking module activation status from JS regularly, it would always return that the module is active (due to the module being force-active).
    * The GA event category should be `${ VIEW_CONTEXT_DASHBOARD }_pagespeed-widget-force-active`.
    * There should be a single tooltip that targets the overall PageSpeed Insights widget.
        * Title: _PageSpeed Insights is now active for everyone_
        * Content: _Keep track of how your pages are doing in terms of speed and user experience and see recommendations on what to improve._

## Implementation Brief
* Using `includes/Modules/PageSpeed_Insights.php`, update `setup_info` to include `'force_active' => true,` in the array returned.
* Using `assets/js/modules/pagespeed-insights/components/dashboard/DashboardPageSpeedWidget.js`, remove the `FallbackComponent` and `IncompleteComponent` options when using the `whenActive` HOC.
* Create `assets/js/feature-tours/page-speed-activation.js` which contains the code for activating the page speed feature tour. Code examples can be taken from other files from the same folder and implemented as per the AC.
* We might need to wait for the PSI widget to load completely before we can activate the feature tour since it takes a long time to load. We can use the same technique used in `assets/js/feature-tours/index.js` for the `allTrafficWidget`, i.e using the `core/ui` data store to store the widget's loading state.

### Test Coverage
* No new tests to be added.
* Existing tests should be updated to cater for above changes.

## QA Brief

* Ensure that on an existing site where you have PageSpeed Insights not connected, it remains disconnected.
    * When you reset the plugin though and then set it up again, it should show up as connected right away.
* Ensure that on a new site PageSpeed Insights is automatically connected without you needing to manually connect it.
* Ensure that in either case it is still always possible to disconnect PageSpeed Insights, as before.
* Whenever PageSpeed Insights is not connected, the blue CTAs to activate PageSpeed Insights which used to appear on the dashboard should no longer be present there.

## Changelog entry

* Update PageSpeed Insights to be active by default for new installs and remove CTAs.

**Comment by felixarntz** [comment]
@tofumatt I moved this back to ACs, since there are a couple more things that need to be specified and included here. First, once PSI is a force-active module, some JS code should be removed, e.g. we no longer need the PSI widget in a CTA state where it prompts to activate the module.

More importantly though, the way it's currently envisioned, this should still happen behind the `unifiedDashboard` feature flag, which is not included in the ACs. However, I am wondering whether that is necessary or whether we could roll out this change individually before - mostly since it would allow a more straightforward solution here, and it seems like a quite decoupled change anyway.

Pinging @marrrmarrr for consent here, let's wait for that decision before moving forward.

**Comment by tofumatt** [comment]
I was wondering that too actually, sounds good 👍🏻

**Comment by marrrmarrr** [comment]
@felixarntz What will the experience be for users with the current dashboard layout? IIRC the tooltips now highlight a specific area of the UI they're pointing to and darken everything else around it. At the same time, PSI widget is near the bottom of the dashboard. 
Does this mean if the user lands on their dashboard, they'll see all content above the fold darkened, but won't see the tooltip unless they scroll down?

**Comment by aaemnnosttv** [comment]
@felixarntz – yes, feature tour tooltips scroll to the focus of the step if needed 👍

**Comment by tofumatt** [comment]
IB ✅

**Comment by ivankruchkoff** [comment]
> It should show (using checkRequirements) if the site does not have pagespeed-insights in the list of active modules from the option. Note that this somehow needs to look at the option value explicitly, since when checking module activation status from JS regularly, it would always return that the module is active (due to the module being force-active).

What about a new selector on `MODULES_PAGESPEED_INSIGHTS` store (as it's going to be active) `isManuallyEnabled` and if that's `false`, it was not enabled previously and so can be the trigger for the tooltip tour in the `checkRequirements()`.

The selector can sit in `assets/js/modules/pagespeed-insights/datastore/manually-activated.js` and call `return API.get( 'modules', 'pagespeed-insights', 'manually-activated'`


On the PHP side, we create a new function in `includes/Core/Modules/Modules.php` `is_manually_enabled( $slug )` which leverages `get_active_modules_option()`.  
https://github.com/google/site-kit-wp/blob/develop/includes/Core/Modules/Modules.php#L990-L1015

An alternative would be to update `get_active_modules()` to take an optional option `$manually_activated_only = false` which would only return modules that are manually activated via the option, avoiding those that are `force_active`
https://github.com/google/site-kit-wp/blob/develop/includes/Core/Modules/Modules.php#L310-L335

We then update the API `includes/Modules/PageSpeed_Insights.php` `create_data_request()` to add a `GET:manually-activated` endpoint to tie the JS request to the PHP processing.

**Comment by tofumatt** [comment]
> What about a new selector on `MODULES_PAGESPEED_INSIGHTS` store (as it's going to be active) `isManuallyEnabled` and if that's `false`, it was not enabled previously and so can be the trigger for the tooltip tour in the `checkRequirements()`.

I like this approach, that works for me. A pedantic note is that I might name the selector `wasManuallyEnabled` or `wasPreviouslyManuallyEnabled` or something, just to signify a status that existed before. I suppose it's still "manually enabled" but as we're evaluating the historical context of the module, the selector referencing the past tense is more intuitive. At least to me! 😅 

But that approach makes sense… I guess the latter approach technically is more flexible in that it allows us to use this approach for future modules that behave the same way, but I'm not convinced that's important or a common enough scenario to build into the `get_active_modules` API. I think it's less complicated to have a separate API for this behaviour, as it's for now unique/one-off and it seems like it would be in the future too.

**Comment by ivankruchkoff** [comment]
Just adding a note that this ends up breaking a lot of 15 of 30 e2e test suites as we have to now factor in `pagespeed-insights` being active everywhere.

**Comment by felixarntz** [comment]
@ivankruchkoff While the selector approach seems cleaner, I'd prefer something simpler. This is a very specific check for something honestly quite odd situation we're in, and even more importantly it is temporary, so we shouldn't over-engineer this. Once this update is out in a release, we can wait for a couple sprints but then this tooltip tour should get removed again, so it's not worth putting too much effort into the infrastructure.

My suggestion here would be to keep it simple and add a new property with whether PSI is manually enabled or not to `_googlesitekitBaseData`.

**Comment by ivankruchkoff** [comment]
Updated to use the `_googlesitekitBaseData`. 
Question for the UX @felixarntz :
If I previously had the `pagespeed-insights` module enabled, which would not trigger the feature tour, and I subsequently reset site kit and set it up again. Should I see the feature tour or not?

As it's currently implemented, the feature tour would trigger, since after the reset, `pagespeed-insights` is no longer manually activated, it is automatically activated, and that causes the feature tour to trigger.

**Comment by felixarntz** [comment]
@ivankruchkoff As briefly discussed earlier, it would be better if the tooltip didn't show after a reset. It makes sense to only show it for users that didn't have the module activated where it then "suddenly" shows up.

A potential way to solve this is could be:
* Include an upgrade routine that sets a very simple option in the DB like `googlesitekit_psi_updated` if the site is currently connected.
* Show the tooltip tour only if that flag is set.
* On reset, delete this option so that it's gone.

**Comment by ivankruchkoff** [comment]
@felixarntz I started working through the CR feedback, and decided to screen record two different scenarios. I think both approaches (now or later) are broken and need some further work to get the ticket ready for QA.

1) Load the feature tour only after the PSI module has finished fetching all data:

It takes about 35s to load, and will provide a weird experience for users who are on the dashboard and are seemingly randomly transported into a feature tour 30s after a page loads. Clock on the right for time reference, shown in real time.

https://user-images.githubusercontent.com/373931/141218137-80cd2b0f-7eac-49cd-8dfe-0960076d8099.mp4

For the first 35s, PSI looks like this:
<img width="1225" alt="Screen Shot 2021-11-10 at 5 04 48 PM" src="https://user-images.githubusercontent.com/373931/141218397-62115450-f225-4231-bbfe-a30b7527a652.png">

2) Load the feature tour as soon as the page is loaded:
Loads instantly, and provides a weird experience as the page expands above the module.


https://user-images.githubusercontent.com/373931/141219502-32b2e2cb-294e-4d08-8465-70d8687c24dd.mp4



Here's a side by side of the page when it first loads, and with everything except PSI loaded:
![before-after-load](https://user-images.githubusercontent.com/373931/141219215-dd73c651-29d4-4f48-bcb8-1154627c4845.png)

We can't predict the height of most popular content table due to word wrap depending on url / title.

**Comment by ivankruchkoff** [comment]
I've addressed the CR feedback hence putting it back into CR, but per my previous comment, I think both approaches (feature tour straight away or feature tour when module data loads) will be problematic.

**Comment by ivankruchkoff** [comment]
Discussed this with @aaemnnosttv to find a compromise approach.

The high level suggestion here for loading the tooltip is to only show it when there is the data available within a very short time period of page load. So it would go as follows:

1. Initial page load, no feature tour as it takes 35s to load the data.
2. Come back to the page at some point, the data is still in the browser cache so it's loaded within half a second, and the feature tour shows up on the page.

The approach is to use the code from:
https://github.com/google/site-kit-wp/pull/4305/commits/6e07da780fa8d2be9e85a194dc55592b879d4b07

Check that we have reports loaded, if they're not, check again via a setTimeout (maximum half second) via a [promise.race](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race) to load the feature tour in a reasonable amount of time rather than at what seems like a random interval.

**Comment by felixarntz** [comment]
@ivankruchkoff @aaemnnosttv The approach you're suggesting is _somewhat_ okay. Preferably, the experience should be that the tooltip is shown right away, even with the widget loading. The real problem there is that then because of the content shift the tooltip tour no longer focuses on the right place - is there no way for us to fix _that_ problem?

I don't think it's a problem at all that the widget itself is still loading while the tooltip tour is shown, since it's not about the widget itself as much as it is about the overall PageSpeed Insights module. With that said, the problem with the focus is definitely a blocker for this approach. If we can find a way to fix it and ensure the tooltip tour stays focused around the same DOM element even when it moves, let's do that. If that's a limitation in the tooltip library that we can't really resolve now, the approach you're suggesting is okay.

**Comment by felixarntz** [comment]
I approved the PR given we go with your suggested approach, but before merging I'd like for us to talk through my response above.

**Comment by aaemnnosttv** [comment]
> The real problem there is that then because of the content shift the tooltip tour no longer focuses on the right place - is there no way for us to fix _that_ problem?

I don't think there is a simple way to fix this at the moment. The only way I think we could guarantee it would be to temporarily fix the height of the other areas. Alternatively, maybe fixing the position of the pagespeed widget area and disabling the scroll on the body might also work but both are pretty hacky.

> If we can find a way to fix it and ensure the tooltip tour stays focused around the same DOM element even when it moves, let's do that. If that's a limitation in the tooltip library that we can't really resolve now, the approach you're suggesting is okay.

@felixarntz Joyride doesn't currently offer a way to enable keeping the target in view while the tooltip is being shown. 

> I approved the PR given we go with your suggested approach, but before merging I'd like for us to talk through my response above.

The PR hasn't been updated with the suggested approach yet, although that should be a fairly small change to make 👍  I've pushed that change for review now.

**Comment by felixarntz** [comment]
Sounds good @aaemnnosttv - I reviewed the additional PR change, lgtm!

One more potential solution here: Would it change if we targeted a selector that only is available once the PSI widget is fully loaded? Maybe doing that would provide the best UX, even if then it doesn't highlight the _full_ widget area. 🤔 What do you think about that @aaemnnosttv @ivankruchkoff?

**Comment by aaemnnosttv** [comment]
> Would it change if we targeted a selector that only is available once the PSI widget is fully loaded?

@felixarntz Sure, but then you risk triggering the tour too late into the page view IMO. Unless a tour is triggered by a specific action, I don't think we should be taking over the screen after the user has had time to start interacting with it. Having your screen hijacked 30s after navigating to the dashboard would not be a good experience. It's a bit more sensitive with this one I think because the area is at the very bottom of the page, where as all other tours have always been in the main viewport right away.

We don't have any feature tours which use in-view as a trigger but this seems like a case where that would be more appropriate rather than "whenever this selector is on the screen".

**Comment by felixarntz** [comment]
@aaemnnosttv Ah right, that's a good point. Let's go with the approach @ivankruchkoff suggested then - I merged the PR.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

A few observations. 

1. The copy on the tooltip `Keep track of how your pages are doing in terms of speed and user experience and see recommendations on what to improve.` doesn't read well IMO.  Maybe it should be `Keep track of how your pages are doing in terms of speed and user experience. See recommendations on what to improve.` Just remove the 'and'. 

2. The tooltip does not appear on small screen sizes like mobile and tablet. Is that correct? 

**Verified:**

- Confirmed you can't disable the pagespeed insights module when on `develop` branch.
- Confirmed that you don't get a feature tour on Site Kit dashboard.

![image](https://user-images.githubusercontent.com/73545194/142017217-ace1bb72-793b-4eb8-b36e-ef0aae3b6420.png)

- Confirmed that the pagespeed insights module is enabled on `develop` branch. 
- Confirmed that you get a feature tour on Site Kit Dashboard

![image](https://user-images.githubusercontent.com/73545194/142017457-48125683-3c05-4b6f-bf9e-b6d97f62e67b.png)

**Comment by ivankruchkoff** [comment]
1. @felixarntz is best suited to answer the question about copy.
2. Yes, correct, see https://github.com/google/site-kit-wp/issues/3003 it applies to all feature tours.
3. Can you show the screenshot of the full window, including the admin toolbar. The tooltip should be fully visible, and if it's not, it's not specific to this ticket but more an issue of the joyride component which handles all of the feature tours, or our way of implementing it.

**Comment by felixarntz** [comment]
@marrrmarrr Can you please chime in on the copy suggestion from @wpdarren in his 1. point in https://github.com/google/site-kit-wp/issues/4133#issuecomment-970428577?

**Comment by marrrmarrr** [comment]
@felixarntz That change SGTM!

**Comment by felixarntz** [comment]
@ivankruchkoff Can you then open a follow-up PR please (against `main`) with that change?

**Comment by ivankruchkoff** [comment]
https://github.com/google/site-kit-wp/pull/4362 @felixarntz

**Comment by felixarntz** [comment]
Merged! Moving back to QA regarding the open question in point 3. in https://github.com/google/site-kit-wp/issues/4133#issuecomment-970558237

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@ivankruchkoff  Re. my point 3 above about the placement of the tooltip when you have no data on the widgets.  This is what the page looks like. Would you like me to create a ticket for this since it appears to be out of scope?

<details>
<summary>Click for Full screen</summary>

![image](https://user-images.githubusercontent.com/73545194/142207822-e8618243-c4ff-42c4-905c-fe53a6084ebc.png)

</details>

@ivankruchkoff @felixarntz I have two additional observations:

1.  I have a new site that does not have PageSpeed activated.  I then change the branch to `develop` or `main` and the module is automatically activated on refresh. When the widget has loaded, the tooltip does not appear.  It's only when you return to the dashboard (or refresh) when the tooltip appears. Is this expected? 

2. When `unifiedDashboard` is activated. I have a new site that does not have PageSpeed activated.  I then change the branch to `develop` or `main` and the module is automatically activated. When the widget has loaded, the tooltip does not get triggered, even if I refresh or go back to the dashboard. At this stage, is this expected?

**Comment by felixarntz** [comment]
@wpdarren Your point 1. is expected based on the discussion above (which started in https://github.com/google/site-kit-wp/issues/4133#issuecomment-965912725).

Point 2. may be indeed broken though - @ivankruchkoff Does the current implementation ensure this tooltip is also shown on the unified dashboard? It can still be relevant there. Kind of a P2, since it would only be relevant if someone later updates from an older version (e.g. 1.45.0) all the way to a version where we've shipped the `unifiedDashboard`, but if it's straightforward we should probably fix it.

**Comment by ivankruchkoff** [comment]
@felixarntz I've put out a change to address the lack of tooltip tour on unified dashboard, https://github.com/google/site-kit-wp/pull/4369 
It will also address the tour being slightly cut off.

Previous selector = `.googlesitekit-widget-area--dashboardSpeed` (red in the next screenshots; it **DOESN'T** appear on unified dashboard )
New selector = `.googlesitekit-widget--pagespeedInsightsWebVitals` (blue in the next screenshots, it **DOES** appear on unified dashboard )

<img width="1330" alt="Screen Shot 2021-11-17 at 12 36 34 PM" src="https://user-images.githubusercontent.com/373931/142279712-a9611f3b-9651-4b72-b4e6-8939e313c0e4.png">


<img width="1313" alt="Screen Shot 2021-11-17 at 12 36 26 PM" src="https://user-images.githubusercontent.com/373931/142279688-86165de2-4d69-4fc0-979e-bd923868d633.png">

**Comment by ivankruchkoff** [comment]
Unified dashboard:
<img width="1440" alt="Screen Shot 2021-11-17 at 12 49 55 PM" src="https://user-images.githubusercontent.com/373931/142280496-8109e325-2b01-4a35-bb8c-89f9ae0a01ea.png">
<img width="1286" alt="Screen Shot 2021-11-17 at 12 50 14 PM" src="https://user-images.githubusercontent.com/373931/142280507-2cef1814-d9d3-4bf7-aaf8-aabe615ba0fc.png">

Pre-unified dashboard:

<img width="1440" alt="Screen Shot 2021-11-17 at 12 49 12 PM" src="https://user-images.githubusercontent.com/373931/142280753-34548e1b-9738-4876-a2f8-c79b909c576b.png">
<img width="1319" alt="Screen Shot 2021-11-17 at 12 49 34 PM" src="https://user-images.githubusercontent.com/373931/142280729-7716d1f6-b96b-4fed-b943-9a1710db1222.png">

However when scrolling back and forth, the following is observed:


https://user-images.githubusercontent.com/373931/142281318-3a2632e7-ac48-4fe8-9df6-f18b320a4b12.mp4

**Comment by felixarntz** [comment]
@ivankruchkoff Thanks for the new PR!

The approach overall sounds solid to me. However, that this now causes another weird bug related to Joyride is unfortunate. Just to confirm, this doesn't happen at all with the current approach correct? In that case, maybe it's better to keep using the widget area selector - we can check if the `unifiedDashoard` feature flag is enabled and if so use the other selector on the tour definition (i.e. the one for the new widget area)?

**Comment by ivankruchkoff** [comment]
@felixarntz sorry that was some overzealous testing on my part I'd removed the call to the promise.race to trigger the displays.

Here's the corrected behaviour on both unified and non-unified.

https://user-images.githubusercontent.com/373931/142284441-3c42d62c-5fe2-4088-8090-ed626c0d3a1f.mp4


https://user-images.githubusercontent.com/373931/142284469-7bcfbc95-94bc-4bf2-8d89-ddeba2795374.mp4

**Comment by wpdarren** [comment]
@ivankruchkoff @felixarntz I've moved this ticket to CR, since it appears to be at this stage right now. Please let me know if that is not the case, and we continue to QA it.

**Comment by felixarntz** [comment]
Providing an update here (partly copied from Slack conversation): The ACs here are largely incorrect due to a misunderstanding between @marrrmarrr and me early on in the definition phase that unfortunately became clear only now (way too late). What the ACs should be is actually a much simpler change:
* The PSI module should be activated by default for new users (or users that reset and then re-setup the plugin). In other words, it should really just be activated per the default value of the option, which would mean existing sites that already have the option set won't be affected.
* The CTA versions of the PSI widget on the dashboard should be removed (since once users consciously deactivate it, we shouldn't prompt them to activate it again).

In other words, most of the more complex changes here (changing to force-enabled, introducing a feature tour, introducing a migration routine for the feature tour) are no longer relevant with the corrected approach here.

We're already very late in this release cycle, so there's two potential paths now:
* Either create a follow-up PR for `main` that includes a bunch of rollbacks from previous commits (e.g. to remove the force-enabled part, the tooltip tour, the related migration class) and the new default value for the active modules option so that PSI is in there by default. In other words, we fix this in time for 1.46.0.
* Or, if the above is too time-consuming for this short time until release, we can revert all work from that issue entirely for this release and then implement the correct change with less time pressure for the next release. In other words, we revert everything to how it was for 1.46.0 and then fix it for 1.47.0.

@aaemnnosttv @ivankruchkoff What do you think is more feasible from a time investment here?

cc @bethanylang

**Comment by aaemnnosttv** [comment]
@felixarntz I was already looking at this so I'll open a quick PR for it, I don't think it should be a problem to include in the release.

**Comment by felixarntz** [comment]
@wpdarren This should be ready for the "new" QA. I've updated the QA Brief (and also ACs) accordingly.

**Comment by wpdarren** [comment]
QA Update: ✅ 

**Verified:**

- On an existing site where PageSpeed Insights not connected, it remains disconnected when I change to `main` branch.
- When I reset the plugin and then set it up again, PSI is connected right away.
- On a new site PageSpeed Insights is automatically connected without needing to manually connect it.
- It is always possible to disconnect PageSpeed Insights, as before.  - [Screenshot](https://user-images.githubusercontent.com/73545194/142612492-5ded6bdc-7584-4c06-b74b-7af4cdf7dcd5.png)
- Whenever PageSpeed Insights is not connected, the blue CTAs to activate PageSpeed Insights are no longer present.
